### ISSUE TITLE
Pyright gets forced into a buggy state when a null character exists in the file.

### ISSUE THREAD

**Issue by ii41** [microsoft/pyright#4559]
**Describe the bug**
When used as a language server, and when there is a null character in the Python file, pyright behaves strangely, and it doesn't restore to a normal state after the character is removed.

For sure, normally the null character shouldn't appear in code, however, vim/neovim has a strange behavior: when a completion item that includes newline characters is inserted, a null character is inserted at the position of each newline character.

https://github.com/vim/vim/issues/2505
https://github.com/neovim/neovim/issues/7769

And this edge case happens to be hit by this strange behavior.

**To Reproduce**
I am unable to copy a piece of code with a null character into the GitHub editor, so I have to ask you to insert the character yourself.

```python
class A:
    def __init__(self) -> None:
        pass
```

Let's use the piece of code above as an example. You can insert a null character at the end of the second line (after `None:`). In vim, that can be done by `CTRL-V CTRL-J` in insert mode, and the inserted null character is shown as `^@`.

After inserting the character, open the file with an editor that is configured to use pyright as a language server. Edit so that the code becomes what's below.

```python
class A:
    def __init__(self) -> None:^@
        pass

    def foo(self) -> None:
        pass
```

At the line of `def foo`, a problem of `Expected indented block` is reported. However, pyright doesn't complain about the invalid character.

Running pyright directly on this file on command line correctly reports the invalid character problem, however.

Now remove the null character. The `Expected indented block` problem continues to be reported.

Continue to do some random editing. The previously reported problem won't disappear, and more strange behaviors will happen, such as reporting problems on valid Python code and providing wrong completion items.

**Expected behavior**
pyright, in its language server mode, should behave normally when and after the null character appears in code.

**VS Code extension or command-line**
Command line version 1.1.291 is used.

**Comment by debonte** [comment]
I'm unable to repro this in VS Code. Here's what I did:

1. Create file with your first code snippet in VS Code and open it with Pyright enabled.
2. Add NUL character in Notepad++ and save it.
3. VS Code picks up external changes and immediately shows the NUL character with an error from Pyright: "Invalid character "\u0" in token" plus the "Unexpected indentation" error, although I see it on the following `pass` line not on the `def foo` line.

![image](https://user-images.githubusercontent.com/12820898/216159528-ff7000e2-bc90-435e-abbd-b9549c8ab44a.png)

If I add the NUL character before opening the file, VS Code complains that the file is binary and even when I force it to open in the text editor, Pyright does not seem to be enabled.

I haven't used neovim before, but I'll give that a shot later this afternoon.

**Comment by erictraut** [comment]
It's possible that the bug is in [coc-pyright](https://github.com/fannheyward/coc-pyright), the neovim LSP plugin for pyright. @fannheyward is the primary author of coc-pyright, so he may be able to help here.

**Comment by ii41** [comment]
I don't use coc-pyright. I use [vim-lsp](https://github.com/prabirshrestha/vim-lsp). This definitely is a pyright problem. I tee'ed the output of pyright into a file, and all the weird behaviors are indeed from the output of pyright.

For reproducing, it is possible that pyright does report the invalid character problem and I overlooked it, but the bigger problem was that after the null gets deleted, the "Unexpected indentation" error does not disappear, and if you continue editing, you'll find more errornous diagnostics and wrong completion. Please finish all the reproducing steps I described, and see whether pyright indeed gets in a problematic state.

**Comment by debonte** [comment]
Should have mentioned that. Yes, removing the NUL character in VS Code causes the indentation error to go away also.

**Comment by ii41** [comment]
Wait... Or it could be that the client doesn't send the "null got deleted" message correctly. If that's the case for sure pyright cannot realize what the actual state of the file is... I overlooked this possibility.

**Comment by ii41** [comment]
Hmmm thank you for trying this out, I'll check how the client handles this. Since vscode does not suffer from this it's more probable that the problem is in vim-lsp. I'm fine with closing this for now.

**Comment by ii41** [comment]
I went to check the vim-lsp's implementation. Indeed, the null character's handling is problematic (or we can say the vimscript API provided by vim/neovim is problematic). Thank you very much for your response. I'll close this and report to vim/neovim or vim-lsp.

**Comment by fannheyward** [comment]
coc.nvim + `coc-pyright` didn't have this issue, two diagnostics reported:

<img width="611" alt="截屏2023-02-02 09 14 08" src="https://user-images.githubusercontent.com/345274/216206784-1118b14d-0dc7-4f40-bdbc-852e434c7957.png">

<img width="475" alt="截屏2023-02-02 09 14 50" src="https://user-images.githubusercontent.com/345274/216206851-a55fcb9c-1608-4c42-bda6-d0276f5fe92c.png">

Same as @debonte's test.

**Comment by ii41** [comment]
@fannheyward vim-lsp uses the vim/neovim builtin `getbufline` function to get buffer contents. This function turns null into newline, and this confuses both the LSP client and the server what the actual content of the buffer is. I reported this to neovim: https://github.com/neovim/neovim/issues/22093. Does coc do anything special to handle this case?
